### ISSUE TITLE
Watching resources without authorisation causes RejectedExecutionException

### ISSUE THREAD

**Issue by adietish** [fabric8io/kubernetes-client#2624]
Steps:
1. ASSERT: have a cluster running where the user isn't authorized (`oc logout`)
1. EXEC: watch a resource type

Result:
An `RejectedExecutionException` is thrown that the user programmer cannot catch because it's thrown by the executor that is running tasks in the `WatchConnectionManager`

```
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@66ac6faf rejected from java.util.concurrent.ScheduledThreadPoolExecutor@626e5c8a[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:326)
	at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:533)
	at java.util.concurrent.ScheduledThreadPoolExecutor.submit(ScheduledThreadPoolExecutor.java:632)
	at java.util.concurrent.Executors$DelegatedExecutorService.submit(Executors.java:678)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.scheduleReconnect(WatchConnectionManager.java:305)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.access$800(WatchConnectionManager.java:50)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onFailure(WatchConnectionManager.java:217)
	at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:571)
	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:198)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by adietish** [comment]
I see the following 2 solutions:
* Allow user programmers to install a `DiscardPolicy` that won't throw (instead of the default `AbortPolicy` which throws and then bubbles up to the user program)
* wait for the reconnect to finish before shutting down and avoid the `RejectedExecutionException`

Thoughts?

**Comment by adietish** [comment]
I see the following 2 solutions:
* Allow user programmers to install a `DiscardPolicy` that won't throw (instead of the default `AbortPolicy` which throws and then bubbles up to the user program)
* `WatchConnectionManager#onFailure` should only try to reconnect if HTTP error codes other than 200 and 503 happen and match the conditions set by `BaseOperation#watch`
* wait for the reconnect to finish before shutting down and avoid the `RejectedExecutionException`

Thoughts?

**Comment by adietish** [comment]
I see the following solutions:
* Dont reconnect in WatchConnectionManager#onFailure
* Allow user programmers to install a `DiscardPolicy` that won't throw (instead of the default `AbortPolicy` which throws and then bubbles up to the user program)
* wait for the reconnect to finish before shutting down and avoid the `RejectedExecutionException` (not great, uncertain when a full stop would happen)

**Comment by adietish** [comment]
This seems fixed in 5.0.0-alpha-2 as in behavior.

**Comment by rtsio** [comment]
I also experience this with #2612.
